### DISCUSSION TITLE
Binding JS variables to python ones in jupyter notebooks

### DISCUSSION THREAD

**Discussion by thorwhalen** [microsoft/vscode-discussions#968] [Extension Development QnA] [answered]
We use the `IPython` library to enable JS-to-python binding (have JS to "talk back" to python) in jupyter notebooks. 
Our code works in jupyter notebooks in the browser, but not in VsCode's jupyter notebooks. 

Here's a simple example to reproduce the problem.

```python
from IPython.display import display, HTML, Javascript

js_code = """
function sendToPython(){
    var data = document.querySelector("#myInput").value;
    var kernel = IPython.notebook.kernel;
    kernel.execute("data_from_js = '" + data + "'");
}

document.querySelector("#myButton").addEventListener("click", function(){
    sendToPython();
});
"""

data_from_js = None

HTML_code = """
<input type="text" id="myInput" placeholder="Enter some text">
<button id="myButton">Submit</button>
<script type="text/Javascript">{}</script>
""".format(js_code)

display(HTML(HTML_code))
```

This will display a text element and a submit button. 
Enter "hello" in input text field, then click "Submit" button. 
Then, in a different cell, if you do `print(data_from_js)`, you should get "hello".

<img width="258" alt="image" src="https://github.com/microsoft/vscode-discussions/assets/1906276/a32d6cf5-9376-4521-ae20-16398f9a295f">

And we do, in a browser-based notebook. 
But in VsCode the `data_from_js` variables remains untouched:

<img width="278" alt="image" src="https://github.com/microsoft/vscode-discussions/assets/1906276/784fe8c3-c322-45d5-9d6c-aa1f809c1201">

The developer tools console says "IPython is not defined":

<img width="474" alt="image" src="https://github.com/microsoft/vscode-discussions/assets/1906276/46d9e89b-2216-4da3-ba41-aa76a6fe453d">

For more details, and attempts at solving this problem, see the [Getting JS to talk back to Python](https://github.com/i2mint/jy/discussions/3) `jy` discussion.

**Comment by gjsjohnmurray** [comment, marked as answer]
Assuming you are not developing an extension, your post doesn't belong here. Please review the pinned README FIRST post at the homepage of this discussion forum.
